File: lab/history.py
"""Git-history mining for the semantic retrieval layer.

Single `git log` pass over the checked-out repo's recent non-merge commit
history. Produces three signals consumed by lanes2.Corpus / lanes2.select_files:

  msgs      - per-file concatenated commit subjects+bodies (a lightweight
              "commit message field" that often names symbols/behavior a
              patch touches even when the file's own source text doesn't),
              capped to each file's MAX_MSGS_PER_FILE most recent commits.
  cochange  - per-file top co-committed partner files with counts (files
              that historically change together are structurally related
              even absent an import edge), PLUS derived "bridge" edges
              between two production files that both co-change with the
              same test-like file. Production files mostly co-change with
              their OWN test file (which select_files' impl_prior excludes
              from the candidate pool), so direct production<->production
              co-change edges are sparse; a shared test file is evidence
              two production files are related even though they never
              appear together in a commit.
  meta      - per-file {"n_commits", "last_ts", "authors"} summary (not yet
              consumed by any scoring path; mined alongside msgs/cochange
              since it's a free byproduct of the same git-log pass).

No network access, no LLM: this is pure `git log` + string processing.

Parsing note: `git log --pretty=format:__C__%at%x00%an%x00%s%n%b --name-only`
emits, per commit: a sentinel-prefixed header line (author unix timestamp,
author name, subject -- NUL-separated), then the raw body (which may itself
contain blank lines between paragraphs), then a git-inserted blank-line
separator, then the changed-file list (one path per line), then another
blank-line separator (or EOF) before the next commit. Blank lines alone can't
distinguish a body paragraph break from the pre-file-list separator, but the
file list is always the LAST blank-line-delimited block before the next
sentinel/EOF, and (unlike prose) every line in it is a bare path with no
spaces -- so we use "last block, if it parses as bare paths" as the file-list
detector and everything before it as the message body. This resolves the
ambiguity without a second git-log pass.
"""

from __future__ import annotations

import subprocess
from collections import Counter, defaultdict
from itertools import combinations
from pathlib import Path

from lanes2 import CODE_EXTENSIONS, _TESTLIKE_RE

_SENTINEL = "__C__"
MAX_MSG_CHARS = 40_000
MAX_MSGS_PER_FILE = 25
BULK_COMMIT_FILE_LIMIT = 20
MIN_COCHANGE_COUNT = 3
MAX_COCHANGE_PARTNERS = 10
MAX_BRIDGE_CANDIDATES = 50  # per test file, perf safety valve (see _bridge_cochange)
MAX_AUTHORS_PER_FILE = 5


def _is_code_file(rel: str) -> bool:
    return rel.endswith(CODE_EXTENSIONS)


def _looks_like_path(ln: str) -> bool:
    s = ln.strip()
    if not s or " " in s or "\t" in s:
        return False
    return "/" in s or ("." in s and not s.startswith("."))


def _split_blocks(lines: list[str]) -> list[list[str]]:
    """Split a list of lines into blank-line-delimited blocks (blanks dropped,
    empty leading/trailing blocks dropped)."""
    blocks: list[list[str]] = []
    cur: list[str] = []
    for ln in lines:
        if ln.strip() == "":
            if cur:
                blocks.append(cur)
                cur = []
        else:
            cur.append(ln)
    if cur:
        blocks.append(cur)
    return blocks


def _parse_commit(subject: str, rest: list[str]) -> tuple[str, list[str]]:
    """rest = every raw line after the header line, up to (excluding) the
    next commit's sentinel line. Returns (message, files)."""
    blocks = _split_blocks(rest)
    if not blocks:
        return subject, []
    last = blocks[-1]
    if all(_looks_like_path(ln) for ln in last):
        files = [ln.strip() for ln in last]
        body_blocks = blocks[:-1]
    else:
        files = []
        body_blocks = blocks
    body = "\n\n".join("\n".join(b) for b in body_blocks)
    message = subject if not body else f"{subject}\n{body}"
    return message, files


def _bridge_cochange(
    cochange_counts: dict[str, Counter[str]],
) -> dict[str, Counter[str]]:
    """Derive production<->production "bridge" edges via a shared test-like
    co-change partner: if impl files a and b both co-change with the same
    test-like file t (each with count >= MIN_COCHANGE_COUNT), add an
    effective a<->b edge with count = min(count_a, count_b) // 2 (kept only
    if >= 2). `cochange_counts` is the raw, untrimmed global co-commit
    counter (already scoped to commits touching <= BULK_COMMIT_FILE_LIMIT
    files, from the caller). Each test file's qualifying partner list is
    capped to MAX_BRIDGE_CANDIDATES (highest-count first) before generating
    pairs, purely to bound combinatorial blowup for hub test files with
    hundreds of historical partners -- since bridge strength is
    min(count_a, count_b), dropping only the lowest-count long-tail
    partners can't remove any pair that would otherwise survive the
    MIN_COCHANGE_COUNT/`>= 2` thresholds ahead of a kept one.
    """
    bridges: dict[str, Counter[str]] = defaultdict(Counter)
    for t, partners in cochange_counts.items():
        if not _TESTLIKE_RE.search(t):
            continue
        qualifying = sorted(
            (
                (f, c) for f, c in partners.items()
                if c >= MIN_COCHANGE_COUNT and not _TESTLIKE_RE.search(f)
            ),
            key=lambda kv: -kv[1],
        )[:MAX_BRIDGE_CANDIDATES]
        for (a, ca), (b, cb) in combinations(qualifying, 2):
            bridge = min(ca, cb) // 2
            if bridge < 2:
                continue
            if bridge > bridges[a][b]:
                bridges[a][b] = bridge
            if bridge > bridges[b][a]:
                bridges[b][a] = bridge
    return bridges


def mine_history(
    repo_path: Path,
    max_commits: int = 5000,
    current_files: set[str] | None = None,
) -> tuple[dict[str, str], dict[str, dict[str, int]], dict[str, dict]]:
    """Mine the last `max_commits` non-merge commits reachable from HEAD.

    Because this runs on the already-checked-out repo (`git log` starts from
    HEAD), commits are naturally bounded to those <= the current checkout's
    commit -- no future leakage relative to whatever base_commit the caller
    checked out beforehand.

    `current_files`, if given, restricts all three return values to files
    that still exist in the current checkout (renamed/deleted paths from
    history are dropped since they can't be retrieval candidates).

    Returns (msgs, cochange, meta):
      msgs[file]     -> str, newest-first concatenated commit messages,
                         capped to MAX_MSGS_PER_FILE commits and MAX_MSG_CHARS.
      cochange[file] -> {other_file: count}, top MAX_COCHANGE_PARTNERS,
                         merging direct co-commit edges and test-bridge edges.
      meta[file]     -> {"n_commits": int, "last_ts": int (unix),
                          "authors": {name: count} top MAX_AUTHORS_PER_FILE}
    """
    if not repo_path.exists():
        return {}, {}, {}

    r = subprocess.run(
        [
            "git", "log", "--no-merges", "-n", str(max_commits),
            f"--pretty=format:{_SENTINEL}%at%x00%an%x00%s%n%b",
            "--name-only",
        ],
        cwd=repo_path, capture_output=True, text=True, timeout=600,
    )
    if r.returncode != 0 or not r.stdout:
        return {}, {}, {}

    lines = r.stdout.splitlines()
    # Locate sentinel-prefixed header lines. A commit subject/body could in
    # principle contain a line that itself starts with the sentinel; guard by
    # only treating a line as a new commit header when we're not already
    # inside an unterminated file-list block (best-effort -- see note above).
    headers: list[int] = [i for i, ln in enumerate(lines) if ln.startswith(_SENTINEL)]

    msgs: dict[str, list[str]] = defaultdict(list)
    cochange_counts: dict[str, Counter[str]] = defaultdict(Counter)
    n_commits: Counter[str] = Counter()
    last_ts: dict[str, int] = {}
    authors: dict[str, Counter[str]] = defaultdict(Counter)
    fileset = current_files

    for idx, start in enumerate(headers):
        end = headers[idx + 1] if idx + 1 < len(headers) else len(lines)
        header = lines[start][len(_SENTINEL):]
        ts_str, _, header_rest = header.partition("\x00")
        author, _, subject = header_rest.partition("\x00")
        try:
            ts = int(ts_str)
        except ValueError:
            ts = 0
        _msg, files = _parse_commit(subject, lines[start + 1: end])
        if fileset is not None:
            files = [f for f in files if f in fileset]
        code_files = [f for f in files if _is_code_file(f)]
        if not code_files:
            continue
        for f in code_files:
            n_commits[f] += 1
            if f not in last_ts:  # git log is newest-first: first seen = most recent
                last_ts[f] = ts
            authors[f][author] += 1
            if len(msgs[f]) < MAX_MSGS_PER_FILE:
                msgs[f].append(_msg)
        if len(files) <= BULK_COMMIT_FILE_LIMIT and len(code_files) >= 2:
            for a, b in combinations(sorted(set(code_files)), 2):
                cochange_counts[a][b] += 1
                cochange_counts[b][a] += 1

    out_msgs: dict[str, str] = {}
    for f, parts in msgs.items():
        # commits are newest-first (git log default order), so this keeps the
        # most recent messages first when truncated to MAX_MSG_CHARS.
        text = "\n".join(parts)
        out_msgs[f] = text[:MAX_MSG_CHARS]

    out_cochange: dict[str, dict[str, int]] = {}
    for f, counter in cochange_counts.items():
        top = [(o, n) for o, n in counter.most_common() if n >= MIN_COCHANGE_COUNT][:MAX_COCHANGE_PARTNERS]
        if top:
            out_cochange[f] = dict(top)

    bridges = _bridge_cochange(cochange_counts)
    for f, bcounter in bridges.items():
        top_bridge = dict(sorted(bcounter.items(), key=lambda kv: -kv[1])[:MAX_COCHANGE_PARTNERS])
        if not top_bridge:
            continue
        merged = dict(out_cochange.get(f, {}))
        for o, c in top_bridge.items():
            merged[o] = max(merged.get(o, 0), c)
        out_cochange[f] = dict(sorted(merged.items(), key=lambda kv: -kv[1])[:MAX_COCHANGE_PARTNERS])

    out_meta: dict[str, dict] = {
        f: {
            "n_commits": n_commits[f],
            "last_ts": last_ts[f],
            "authors": dict(authors[f].most_common(MAX_AUTHORS_PER_FILE)),
        }
        for f in n_commits
    }

    return out_msgs, out_cochange, out_meta

File: bgrep-rs/src/history.rs
//! Git-history mining for the semantic retrieval layer -- a faithful port of
//! `lab/history.py`. Single `git log` pass over the checked-out repo's recent
//! non-merge commit history, producing `msgs` (per-file commit message
//! field), `cochange` (per-file co-committed partner counts, plus derived
//! test-bridge edges), and `meta` (per-file commit/author summary, not
//! consumed by any scoring path -- kept for parity with the Python return
//! shape only).
//!
//! ORDERING NOTE: unlike select_files() in core.rs (which has exactly one
//! raw-`set`-iteration nondeterminism -- see PARITY_NOTES.md), history.py's
//! own logic never iterates a raw Python `set` in an order-sensitive way:
//! every dict here is insertion-ordered (deterministic, driven by git log's
//! own -- deterministic -- commit order), and the one spot combinations()
//! runs over a set (`combinations(sorted(set(code_files)), 2)`) is
//! explicitly pre-sorted first. So `IndexMap`, used throughout below to
//! mirror Python dict/Counter insertion order, gives byte-for-byte parity
//! with no caveats.

use crate::core::{is_code_file, TESTLIKE_RE};
use indexmap::IndexMap;
use serde::{Deserialize, Serialize};
use std::collections::HashSet;
use std::path::Path;
use std::process::Command;

const SENTINEL: &str = "__C__";
const MAX_MSG_CHARS: usize = 40_000;
const MAX_MSGS_PER_FILE: usize = 25;
const BULK_COMMIT_FILE_LIMIT: usize = 20;
const MIN_COCHANGE_COUNT: i64 = 3;
const MAX_COCHANGE_PARTNERS: usize = 10;
const MAX_BRIDGE_CANDIDATES: usize = 50;
const MAX_AUTHORS_PER_FILE: usize = 5;

/// Insertion-ordered multiset, mirroring `collections.Counter`'s dict
/// semantics (first-seen order preserved for ties in `most_common()`).
pub type OrderedCounter = IndexMap<String, i64>;

fn counter_incr(c: &mut OrderedCounter, k: &str, by: i64) {
    *c.entry(k.to_string()).or_insert(0) += by;
}

/// `Counter.most_common()`: stable sort by count descending, ties broken by
/// the counter's insertion order (Rust's `sort_by` is stable, and the
/// intermediate Vec is built by iterating the IndexMap in insertion order,
/// so this matches exactly).
fn most_common(c: &OrderedCounter) -> Vec<(String, i64)> {
    let mut v: Vec<(String, i64)> = c.iter().map(|(k, n)| (k.clone(), *n)).collect();
    v.sort_by(|a, b| b.1.cmp(&a.1));
    v
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct FileMeta {
    pub n_commits: i64,
    pub last_ts: i64,
    pub authors: IndexMap<String, i64>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct HistoryData {
    pub msgs: IndexMap<String, String>,
    pub cochange: IndexMap<String, IndexMap<String, i64>>,
    pub meta: IndexMap<String, FileMeta>,
}

fn looks_like_path(ln: &str) -> bool {
    let s = ln.trim();
    if s.is_empty() || s.contains(' ') || s.contains('\t') {
        return false;
    }
    s.contains('/') || (s.contains('.') && !s.starts_with('.'))
}

/// Split a list of lines into blank-line-delimited blocks (blanks dropped,
/// empty leading/trailing blocks dropped).
fn split_blocks<'a>(lines: &[&'a str]) -> Vec<Vec<&'a str>> {
    let mut blocks = Vec::new();
    let mut cur: Vec<&str> = Vec::new();
    for &ln in lines {
        if ln.trim().is_empty() {
            if !cur.is_empty() {
                blocks.push(std::mem::take(&mut cur));
            }
        } else {
            cur.push(ln);
        }
    }
    if !cur.is_empty() {
        blocks.push(cur);
    }
    blocks
}

/// `rest` = every raw line after the header line, up to (excluding) the next
/// commit's sentinel line. Returns (message, files).
fn parse_commit(subject: &str, rest: &[&str]) -> (String, Vec<String>) {
    let blocks = split_blocks(rest);
    if blocks.is_empty() {
        return (subject.to_string(), Vec::new());
    }
    let last = blocks.last().unwrap();
    let (files, body_blocks): (Vec<String>, &[Vec<&str>]) = if last.iter().all(|ln| looks_like_path(ln)) {
        (
            last.iter().map(|ln| ln.trim().to_string()).collect(),
            &blocks[..blocks.len() - 1],
        )
    } else {
        (Vec::new(), &blocks[..])
    };
    let body = body_blocks
        .iter()
        .map(|b| b.join("\n"))
        .collect::<Vec<_>>()
        .join("\n\n");
    let message = if body.is_empty() {
        subject.to_string()
    } else {
        format!("{subject}\n{body}")
    };
    (message, files)
}

/// Derive production<->production "bridge" edges via a shared test-like
/// co-change partner (see history.py's `_bridge_cochange` docstring for the
/// full rationale).
fn bridge_cochange(cochange_counts: &IndexMap<String, OrderedCounter>) -> IndexMap<String, OrderedCounter> {
    let mut bridges: IndexMap<String, OrderedCounter> = IndexMap::new();
    for (t, partners) in cochange_counts {
        if !TESTLIKE_RE.is_match(t) {
            continue;
        }
        let mut qualifying: Vec<(String, i64)> = partners
            .iter()
            .filter(|(f, c)| **c >= MIN_COCHANGE_COUNT && !TESTLIKE_RE.is_match(f))
            .map(|(f, c)| (f.clone(), *c))
            .collect();
        qualifying.sort_by(|a, b| b.1.cmp(&a.1));
        qualifying.truncate(MAX_BRIDGE_CANDIDATES);

        for i in 0..qualifying.len() {
            for j in (i + 1)..qualifying.len() {
                let (a, ca) = &qualifying[i];
                let (b, cb) = &qualifying[j];
                let bridge = ca.min(cb) / 2;
                if bridge < 2 {
                    continue;
                }
                let cur_ab = bridges.entry(a.clone()).or_default().get(b).copied().unwrap_or(0);
                if bridge > cur_ab {
                    bridges.entry(a.clone()).or_default().insert(b.clone(), bridge);
                }
                let cur_ba = bridges.entry(b.clone()).or_default().get(a).copied().unwrap_or(0);
                if bridge > cur_ba {
                    bridges.entry(b.clone()).or_default().insert(a.clone(), bridge);
                }
            }
        }
    }
    bridges
}

/// Mine the last `max_commits` non-merge commits reachable from HEAD.
pub fn mine_history(
    repo_path: &Path,
    max_commits: usize,
    current_files: Option<&HashSet<String>>,
) -> HistoryData {
    if !repo_path.exists() {
        return HistoryData::default();
    }

    let pretty = format!("--pretty=format:{SENTINEL}%at%x00%an%x00%s%n%b");
    let output = Command::new("git")
        .args([
            "log",
            "--no-merges",
            "-n",
            &max_commits.to_string(),
            &pretty,
            "--name-only",
        ])
        .current_dir(repo_path)
        .output();
    let output = match output {
        Ok(o) if o.status.success() => o,
        _ => return HistoryData::default(),
    };
    let stdout = String::from_utf8_lossy(&output.stdout).into_owned();
    if stdout.is_empty() {
        return HistoryData::default();
    }

    let lines: Vec<&str> = crate::pyutil::py_splitlines(&stdout);
    let headers: Vec<usize> = lines
        .iter()
        .enumerate()
        .filter(|(_, ln)| ln.starts_with(SENTINEL))
        .map(|(i, _)| i)
        .collect();

    let mut msgs: IndexMap<String, Vec<String>> = IndexMap::new();
    let mut cochange_counts: IndexMap<String, OrderedCounter> = IndexMap::new();
    let mut n_commits: OrderedCounter = IndexMap::new();
    let mut last_ts: IndexMap<String, i64> = IndexMap::new();
    let mut authors: IndexMap<String, OrderedCounter> = IndexMap::new();

    for (idx, &start) in headers.iter().enumerate() {
        let end = headers.get(idx + 1).copied().unwrap_or(lines.len());
        let header = &lines[start][SENTINEL.len()..];
        let mut parts = header.splitn(3, '\u{0}');
        let ts_str = parts.next().unwrap_or("");
        let author = parts.next().unwrap_or("");
        let subject = parts.next().unwrap_or("");
        let ts: i64 = ts_str.parse().unwrap_or(0);

        let (_msg, mut files) = parse_commit(subject, &lines[start + 1..end]);
        if let Some(fs) = current_files {
            files.retain(|f| fs.contains(f));
        }
        let n_files_total = files.len(); // pre-code-filter count, matches Python's `len(files)`
        let code_files: Vec<String> = files.into_iter().filter(|f| is_code_file(f)).collect();
        if code_files.is_empty() {
            continue;
        }
        for f in &code_files {
            counter_incr(&mut n_commits, f, 1);
            last_ts.entry(f.clone()).or_insert(ts);
            counter_incr(authors.entry(f.clone()).or_default(), author, 1);
            let list = msgs.entry(f.clone()).or_default();
            if list.len() < MAX_MSGS_PER_FILE {
                list.push(_msg.clone());
            }
        }
        if n_files_total <= BULK_COMMIT_FILE_LIMIT {
            let mut uniq: Vec<&String> = code_files.iter().collect::<HashSet<_>>().into_iter().collect();
            uniq.sort();
            if uniq.len() >= 2 {
                for i in 0..uniq.len() {
                    for j in (i + 1)..uniq.len() {
                        let a = uniq[i].clone();
                        let b = uniq[j].clone();
                        counter_incr(cochange_counts.entry(a.clone()).or_default(), &b, 1);
                        counter_incr(cochange_counts.entry(b).or_default(), &a, 1);
                    }
                }
            }
        }
    }

    let mut out_msgs: IndexMap<String, String> = IndexMap::new();
    for (f, parts) in &msgs {
        let text = parts.join("\n");
        let truncated: String = text.chars().take(MAX_MSG_CHARS).collect();
        out_msgs.insert(f.clone(), truncated);
    }

    let mut out_cochange: IndexMap<String, IndexMap<String, i64>> = IndexMap::new();
    for (f, counter) in &cochange_counts {
        let top: Vec<(String, i64)> = most_common(counter)
            .into_iter()
            .filter(|(_, n)| *n >= MIN_COCHANGE_COUNT)
            .take(MAX_COCHANGE_PARTNERS)
            .collect();
        if !top.is_empty() {
            let m: IndexMap<String, i64> = top.into_iter().collect();
            out_cochange.insert(f.clone(), m);
        }
    }

    let bridges = bridge_cochange(&cochange_counts);
    for (f, bcounter) in &bridges {
        let mut top_bridge_v = most_common(bcounter);
        top_bridge_v.truncate(MAX_COCHANGE_PARTNERS);
        if top_bridge_v.is_empty() {
            continue;
        }
        let top_bridge: IndexMap<String, i64> = top_bridge_v.into_iter().collect();
        let mut merged: IndexMap<String, i64> = out_cochange.get(f).cloned().unwrap_or_default();
        for (o, c) in &top_bridge {
            let cur = merged.get(o).copied().unwrap_or(0);
            merged.insert(o.clone(), cur.max(*c));
        }
        let mut merged_v: Vec<(String, i64)> = merged.into_iter().collect();
        merged_v.sort_by(|a, b| b.1.cmp(&a.1));
        merged_v.truncate(MAX_COCHANGE_PARTNERS);
        out_cochange.insert(f.clone(), merged_v.into_iter().collect());
    }

    let mut out_meta: IndexMap<String, FileMeta> = IndexMap::new();
    for (f, &n) in &n_commits {
        let mut auth_v = most_common(authors.get(f).unwrap());
        auth_v.truncate(MAX_AUTHORS_PER_FILE);
        out_meta.insert(
            f.clone(),
            FileMeta {
                n_commits: n,
                last_ts: *last_ts.get(f).unwrap_or(&0),
                authors: auth_v.into_iter().collect(),
            },
        );
    }

    HistoryData {
        msgs: out_msgs,
        cochange: out_cochange,
        meta: out_meta,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn looks_like_path_cases() {
        assert!(looks_like_path("a/b.py"));
        assert!(looks_like_path("foo.py"));
        assert!(!looks_like_path(".hidden"));
        assert!(!looks_like_path("has space"));
        assert!(!looks_like_path(""));
        assert!(!looks_like_path("plainword"));
    }

    #[test]
    fn parse_commit_splits_file_list() {
        let rest = vec!["body line 1", "", "a/b.py", "c/d.py"];
        let (msg, files) = parse_commit("subject", &rest);
        assert_eq!(msg, "subject\nbody line 1");
        assert_eq!(files, vec!["a/b.py", "c/d.py"]);
    }

    #[test]
    fn parse_commit_no_file_list_block() {
        // last block doesn't parse as bare paths -> treated as body
        let rest = vec!["not a path list here"];
        let (msg, files) = parse_commit("subject", &rest);
        assert_eq!(msg, "subject\nnot a path list here");
        assert!(files.is_empty());
    }
}
